runner/src/main.rs

```diff
@@ -0,0 +1,122 @@
+
+    let cmdline_matches = clap::App::new("runner")
+        .arg(clap::Arg::with_name("delay").help("Reset delay in milliseconds")
+             .long("delay").short("d").takes_value(true))
+        .arg(clap::Arg::with_name("test").long("test").short("t"))
+        .get_matches();
+
+    // Parse the command line arguments early so that we fail fast (with a nice
+    // error message) if we cannot parse them. This avoids resetting the H1B if
+    // a bad command line argument is used.
+    let delay = cmdline_matches.value_of("delay")
+        .map_or(100, |d| d.parse().expect("Unable to parse --delay value"));
+
+    // When this runner starts, the H1B will already be running. As a result, we
+    // may have missed some of its output. This is particularly problematic for
+    // --test, as we may have missed important markers.
+    //
+    // To collect as much of the H1B's output as possible, we perform the
+    // following sequence:
+    //   1. Power down the H1B (write "0").
+    //   2. Wait a bit (unfortunately we don't have a way to know for sure
+    //      whether the device is powered down). Flush the target's console
+    //      during/after the wait.
+    //   3. Start listening to the debug console output (this happens
+    //      implicitly).
+    //   4. Power up the H1B (write "1").
+    let mut debug_console = std::fs::OpenOptions::new()
+                            .append(true)
+                            .open("/dev/ttyUltraConsole3")
+                            .expect("Unable to open /dev/ttyUltraConsole3");
+    // 1. Power down the H1B
+    debug_console.write_all(b"0").expect("Unable to reset H1B (failed write)");
+    debug_console.flush().expect("Unable to reset H1B (failed flush)");
+
+    // 2. Wait for --delay milliseconds.
+    std::thread::sleep(std::time::Duration::from_millis(delay));
+
+    // 3. Open the console
+    let target_console = std::fs::OpenOptions::new()
+                         .read(true)
+                         .open("/dev/ttyUltraTarget2")
+                         .expect("Unable to open /dev/ttyUltraTarget2");
+
+    // 4. Power up the H1B.
+    debug_console.write_all(b"1").expect("Unable to restart H1B (failed write)");
+    debug_console.flush().expect("Unable to restart H1B (failed flush)");
+
+    // If we're not in --test mode, return 0 on SIGINT.
+    let test_mode = cmdline_matches.is_present("test");
+    if !test_mode {
+        unsafe { libc::signal(libc::SIGINT, sigint_handler as usize); }
+    }
+
+    // Stream in the console output, and echo it to stdout. If --test was
+    // passed, we search for \nTEST_FINISHED: [FAIL|SUCCESS]\n and terminate
+    // (with the corresponding error code) once found.
+    let fail_message = b"\nTEST_FINISHED: FAIL\n";
+    let success_message = b"\nTEST_FINISHED: SUCCESS\n";
+    // The buffer length needs to match the larger of fail_message and
+    // success_message.
+    let mut buffer = [0; 24];
```

Comment:
If it needs to match the larger of the two, why not compute it? If you ever change one of those messages to be significantly longer this will break.